torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -408,35 +444,45 @@ void TensorPipeAgent::onListenerAccepted(
 
 void TensorPipeAgent::pipeRead(
     const std::shared_ptr<tensorpipe::Pipe>& pipe,
-    std::function<void(const tensorpipe::Error&, Message&&)> fn) noexcept {
-  pipe->readDescriptor([fn{std::move(fn)}, pipe](
+    std::function<void(
+        const tensorpipe::Error&,
+        Message&&,
+        std::shared_ptr<FullDeviceContext>)> fn) noexcept {
+  pipe->readDescriptor([fn{std::move(fn)}, pipe, this](
                            const tensorpipe::Error& error,
                            tensorpipe::Message tpMessage) mutable {
+    // TODO: pass these streams to TensorPipe when it can accept streams
+    auto ctx = createFullDeviceContext(
+        reverseDeviceMaps_.empty() && opts_.deviceMaps.empty());
+
     if (error) {
-      fn(error, Message());
+      fn(error, Message(), std::move(ctx));
       return;
     }
 
-    TensorpipeReadBuffers tpBuffers = tensorpipeAllocate(tpMessage);
+    TensorpipeReadBuffers tpBuffers = tensorpipeAllocate(tpMessage, ctx);
 
     pipe->read(
         std::move(tpMessage),
         [tpBuffers{
-             std::make_shared<TensorpipeReadBuffers>(std::move(tpBuffers))},
-         fn{std::move(fn)}](
+            std::make_shared<TensorpipeReadBuffers>(std::move(tpBuffers))},
+         fn{std::move(fn)},
+         ctx{std::move(ctx)}](
             const tensorpipe::Error& error,
             tensorpipe::Message tpMessage) mutable {
           if (error) {
-            fn(error, Message());
+            fn(error, Message(), std::move(ctx));
             return;
           }
 
           // FIXME This does some unpickling, which could be a bit expensive:
           // perhaps it would be best to perform it inside the worker threads?
+          ctx->blockCurrentStreams();
           Message rpcMessage = tensorpipeDeserialize(
               std::move(tpMessage), std::move(*tpBuffers));
 
-          fn(error, std::move(rpcMessage));
+          ctx->recordDataPtrs(tpBuffers->tensors);
```

Comment:
Shouldn't this be done before `tensorpipeDeserialize`? Also, we moved tpBuffers into `tensorpipeDeserialize` above and as a result we might have an empty/null object here?